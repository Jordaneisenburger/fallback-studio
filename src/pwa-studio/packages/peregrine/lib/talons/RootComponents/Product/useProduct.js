import { useMemo } from 'react';
import { useApolloClient, useQuery } from '@apollo/react-hooks';

/**
 * A [React Hook]{@link https://reactjs.org/docs/hooks-intro.html} that
 * controls the logic for the Product Root Component.
 *
 * @kind function
 *
 * @param {object}      props
 * @param {String}      props.cachePrefix - The prefix to apply to the cache key.
 * @param {GraphQLAST}  props.fragment - The GraphQL fragment to match against a cache entry.
 * @param {Function}    props.mapProduct - A function for updating products to the proper shape.
 * @param {GraphQLAST}  props.query - The query to fetch a product.
 * @param {String}      props.urlKey - The url_key of this product.
 *
 * @returns {object}    result
 * @returns {Bool}      result.error - Indicates a network error occurred.
 * @returns {Bool}      result.loading - Indicates the query is in flight.
 * @returns {Bool}      result.product - The product's details.
 */
export const useProduct = props => {
    const { cachePrefix, fragment, mapProduct, query, urlKey } = props;

    const apolloClient = useApolloClient();

    /*
     * Look up the product in cache first.
     *
     * We may have it from a previous query, but we have to manually tell Apollo cache where to find it.
     * A single product query (as described by https://github.com/magento/graphql-ce/issues/86)
     * will alleviate the need to do this manual work.
     *
     * If the object with the specified `id` is not in the cache, we get `null`.
     * If the object is in the cache but doesn't have all the fields we need, an error is thrown.
     *
     * @see https://www.apollographql.com/docs/react/caching/cache-interaction/#readfragment.
     */
    const productFromCache = useMemo(() => {
        try {
            return apolloClient.readFragment({
                id: `${cachePrefix}:${urlKey}`,
                fragment,
                variables: { onServer: false }
            });
        } catch (e) {
            // The product is in the cache but it is missing some fields the fragment needs.
            return null;
        }
    }, [apolloClient, cachePrefix, fragment, urlKey]);

    /*
     * Fetch the product from the network.
     *
     * Note that this always fires, regardless of whether we have a cached product or not:
     * If we do not have a cached product, we need to go fetch it from the network.
     * If we do have a cached product, we want to ensure its cache entry doesn't get stale.
     * In both cases, Apollo will update the cache with the latest data when this returns.
     */
    const { loading, error, data } = useQuery(query, {
        // Once we're able to remove the manual cache lookup,
        // this fetch policy can change to 'cache-and-network'.
        fetchPolicy: 'network-only',
        variables: {
            onServer: false,
            urlKey
        }
    });

    const product = useMemo(() => {
        if (productFromCache) {
            return mapProduct(productFromCache);
        }

        if (data) {
            const productFromNetwork = data.productDetail.items[0];
            return mapProduct(productFromNetwork);
        }

        // The product isn't in the cache and we don't have a response from GraphQL yet.
        return null;
    }, [data, mapProduct, productFromCache]);

    return {
        error,
        loading,
        product
    };
};
