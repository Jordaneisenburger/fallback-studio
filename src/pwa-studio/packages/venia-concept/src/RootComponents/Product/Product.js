import React, { Component } from 'react';
import { string, func } from 'prop-types';

import { connect, Query } from 'src/drivers';
import { addItemToCart } from 'src/actions/cart';
import { loadingIndicator } from 'src/components/LoadingIndicator';
import ProductFullDetail from 'src/components/ProductFullDetail';
import getUrlKey from 'src/util/getUrlKey';
import productQuery from 'src/queries/getProductDetail.graphql';

/**
 * As of this writing, there is no single Product query type in the M2.3 schema.
 * The recommended solution is to use filter criteria on a Products query.
 * However, the `id` argument is not supported. See
 * https://github.com/magento/graphql-ce/issues/86
 * TODO: Replace with a single product query when possible.
 */
class Product extends Component {
    static propTypes = {
        addItemToCart: func.isRequired,
        cartId: string
    };

    addToCart = async (item, quantity) => {
        const { addItemToCart, cartId } = this.props;
        await addItemToCart({ cartId, item, quantity });
    };

    componentDidMount() {
        window.scrollTo(0, 0);
    }

    // map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to maintain backwards compatibility
    mapProduct(product) {
        const { description } = product;
        return {
            ...product,
            description:
                typeof description === 'object' ? description.html : description
        };
    }

    render() {
        return (
            <Query
                query={productQuery}
                variables={{ urlKey: getUrlKey(), onServer: false }}
            >
                {({ loading, error, data }) => {
                    if (error) return <div>Data Fetch Error</div>;
                    if (loading) return loadingIndicator;

                    const product = data.productDetail.items[0];

                    return (
                        <ProductFullDetail
                            product={this.mapProduct(product)}
                            addToCart={this.props.addItemToCart}
                        />
                    );
                }}
            </Query>
        );
    }
}

const mapDispatchToProps = {
    addItemToCart
};

export default connect(
    null,
    mapDispatchToProps
)(Product);
