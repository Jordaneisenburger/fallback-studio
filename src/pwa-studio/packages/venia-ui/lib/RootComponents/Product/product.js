import React, { Fragment, useEffect, useMemo } from 'react';
import { useQuery } from '@apollo/react-hooks';
import { Title } from '../../components/Head';
import ErrorView from '../../components/ErrorView';
import { fullPageLoadingIndicator } from '../../components/LoadingIndicator';
import ProductFullDetail from '../../components/ProductFullDetail';
import getUrlKey from '../../util/getUrlKey';

/*
 * As of this writing, there is no single Product query type in the M2.3 schema.
 * The recommended solution is to use filter criteria on a Products query.
 * However, the `id` argument is not supported. See
 * https://github.com/magento/graphql-ce/issues/86
 * TODO: Replace with a single product query when possible.
 */
import GET_PRODUCT_DETAIL from '../../queries/getProductDetail.graphql';

const Product = () => {
    useEffect(() => {
        window.scrollTo(0, 0);
    }, []);

    const { loading, error, data } = useQuery(GET_PRODUCT_DETAIL, {
        variables: {
            onServer: false,
            urlKey: getUrlKey()
        }
    });

    // Memoize the result from the query to avoid unnecessary rerenders.
    const product = useMemo(() => {
        if (!data) {
            return;
        }
        const product = data.productDetail.items[0];
        // map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to
        // maintain backwards compatibility
        const { description } = product;
        return {
            ...product,
            description:
                typeof description === 'object' ? description.html : description
        };
    }, [data]);

    if (loading) return fullPageLoadingIndicator;
    if (error) return <div>Data Fetch Error</div>;

    if (!product) {
        return <ErrorView outOfStock={true} />;
    }

    // Note: STORE_NAME is injected by Webpack at build time.
    return (
        <Fragment>
            <Title>{`${product.name} - ${STORE_NAME}`}</Title>
            <ProductFullDetail product={product} />
        </Fragment>
    );
};

export default Product;
