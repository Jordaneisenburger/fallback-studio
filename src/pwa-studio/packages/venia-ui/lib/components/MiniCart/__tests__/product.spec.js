import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Product from '../product';

global.getComputedStyle = jest.fn().mockReturnValue({
    getPropertyValue: jest.fn().mockReturnValue('80px')
});
jest.mock('react', () => {
    const React = jest.requireActual('react');
    const memoSpy = jest.spyOn(React, 'useMemo');
    const stateSpy = jest.spyOn(React, 'useState');

    return Object.assign(React, {
        useMemo: memoSpy,
        useState: stateSpy
    });
});

const renderer = new ShallowRenderer();

const props = {
    beginEditItem: jest.fn(),
    currencyCode: 'US',
    item: {
        image: {
            file: 'unittest'
        },
        name: 'Unit Test Product',
        options: [],
        price: 99,
        qty: 1
    },
    removeItemFromCart: jest.fn()
};

test('it renders correctly', () => {
    const tree = renderer.render(<Product {...props} />);

    expect(tree).toMatchSnapshot();
});
