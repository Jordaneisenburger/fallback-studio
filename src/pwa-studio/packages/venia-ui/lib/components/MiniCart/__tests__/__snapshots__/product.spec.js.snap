// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<li>
  <Image
    alt="Unit Test Product"
    classes={
      Object {
        "root": undefined,
      }
    }
    fileSrc="unittest"
    placeholder="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII="
    sizes="80px"
    src="/media/catalog/product/unittest?auto=webp&format=pjpg&width=80&height=100"
  />
  <div>
    Unit Test Product
  </div>
  <ProductOptions
    options={Array []}
  />
  <div>
    <div>
      <span>
        1
      </span>
      <span>
        Ã—
      </span>
      <span>
        <mockConstructor
          currencyCode="US"
          value={99}
        />
      </span>
    </div>
  </div>
  <Kebab>
    <Section
      icon="Heart"
      isFilled={false}
      onClick={[Function]}
      text="Add to favorites"
    />
    <Section
      icon="Edit2"
      onClick={[Function]}
      text="Edit item"
    />
    <Section
      icon="Trash"
      onClick={[Function]}
      text="Remove item"
    />
  </Kebab>
</li>
`;
